### PR TITLE
user-defined types ter: clean, refactor basic_json class

### PR DESCRIPTION
**Pull request by theodelrieu**
Hi!

This PR is the follow-up of #355 

I managed to replace a lot of constructors with `from/to_json` overloads, which helped me to remove some SFINAE boilerplate, and to add a new feature: giving full control on type serialization to whom needs it.

I advise reviewers to review commit per commit, I tried my best to separate everything.

# Things that need to be done
- Put back and adapt the docs (I got a heavy hand on the 'delete' button)
- static assert message improvement
- improve README (talk about specializing types supported by the library, improve clarity)
- TODO fill this row if something comes to mind

### CODE REVIEW

File: src/json.hpp [hunk 1]
```diff
@@ -3255,11 +3268,13 @@ class basic_json
...
+              enable_if_t<not std::is_pointer<ValueType>::value and
+                              detail::is_compatible_basic_json_type<
+                                  uncvref_t<ValueType>, basic_json_t>::value,
+                          int> = 0>
     auto get() const
-    -> decltype(this->get_impl(static_cast<ValueType*>(nullptr)))
-    {
-        return get_impl(static_cast<ValueType*>(nullptr));
+        -> decltype(this->get_impl(static_cast<ValueType *>(nullptr))) {
+      return get_impl(static_cast<ValueType *>(nullptr));
```
Comment (gregmarr) [inline src/json.hpp]: The brace here doesn't match the style of the rest of the library.

File: src/json.hpp [hunk 2]
```diff
@@ -424,6 +440,22 @@ struct has_to_json
...
+{
+  external_constructor<value_t::boolean>::construct(j, b);
+}
+
+template <typename Json>
+void from_json(Json const& j, typename Json::boolean_t& b)
+{
+  if (!j.is_boolean())
+    throw std::domain_error("type must be boolean, but is " + type_name(j));
+  b = *const_cast<Json&>(j).template get_ptr<typename Json::boolean_t*>();
```
Comment (gregmarr) [inline src/json.hpp]: Is there no way to do this without a const_cast?

File: src/json.hpp [hunk 3]
```diff
@@ -456,6 +478,14 @@ void from_json(Json const& j, typename Json::boolean_t& b)
   b = *const_cast<Json&>(j).template get_ptr<typename Json::boolean_t*>();
 }
 
+template <typename Json>
+void from_json(Json const& j, typename Json::string_t& s)
+{
+  if (!j.is_string())
+    throw std::domain_error("type must be string, but is " + type_name(j));
+  s = *const_cast<Json&>(j).template get_ptr<typename Json::string_t*>();
```
Comment (gregmarr) [inline src/json.hpp]: As in boolean version, can this be done without const_cast?

File: src/json.hpp [hunk 4]
```diff
@@ -11398,9 +11344,21 @@ class basic_json
     }
 
     private:
+    friend bool operator==(json_pointer const &lhs,
+                           json_pointer const &rhs) noexcept
+    {
+      return lhs.reference_tokens == rhs.reference_tokens;
```
Comment (gregmarr) [inline src/json.hpp]: Looks like this library uses 4 space indents.

File: src/json.hpp [hunk 5]
```diff
@@ -11398,9 +11344,21 @@ class basic_json
...
+    friend bool operator!=(json_pointer const &lhs,
+                           json_pointer const &rhs) noexcept
+    {
+      return !(lhs == rhs);
+    }
+
     /// the reference tokens
     std::vector<std::string> reference_tokens {};
-               };
+    };
```
Comment (gregmarr) [inline src/json.hpp]: Something still seems off about this brace indentation.

File: src/json.hpp (hunk 2 above)
Comment (theodelrieu) [inline src/json.hpp]: I think it can, adding const to the `get_ptr` template arg should be good

File: src/json.hpp (hunk 4 above)
Comment (theodelrieu) [inline src/json.hpp]: I added `run make pretty to the list of things to be done

**Comment by theodelrieu** [comment]
I managed to fix the static_assert for one method only on MSVC.

For a yet unknown reason, doing the exact same thing than `has_from_json_adl` with to_json fails miserably, with MSVC trying to convert ints to `std::ios_base::iostate` and other funky things.

I will commit the failing code tomorrow, hoping that somebody will find a workaround..

**Comment by theodelrieu** [comment]
I think the `unit-udt.cpp` tests are quite good. And the whole library uses the `to/from_json` machinery now, that adds even more testing to the feature.

If someone finds an edge-case, or something I missed, please let me know

**Comment by theodelrieu** [comment]
@nlohmann Hi Niels, do you think we could meet on Slack soon? I'd like to have your input on some things 
(mainly documentation, the refactoring and some questions that are hidden in code comments right now)

See you soon!

**Comment by theodelrieu** [comment]
I added a bit of doc, but frankly I'm terrible at this, so I'd gladly take some help on this part!

But once the doc and examples are done, I think we can merge it.

File: README.md [hunk 6]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+j["address"] = p.address;
+j["age"] = p.age;
+
+// ...
+
+// convert from JSON
+ns::person p {j["name"].get<std::string>(), j["address"].get<std::string>(), j["age"].get<int>()};
+```
+
+It works, but that's quite a lot of boilerplate.. Hopefully, there's a better way:
```
Comment (nlohmann) [inline README.md]: typo: `..`

Comment (nlohmann) [inline README.md]: I would replace "Hopefully" with "Fortunately".

File: README.md [hunk 7]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+}
+} // namespace ns
+```
+
+That's all. When calling the json constructor with your type, your custom `to_json` method will be automatically called.
+Likewise, when calling `get<your_type>()`, the `from_json` method will be called.
+
+Some important things:
+
+* Those methods **MUST** be in your type's namespace, or the library will not be able to locate them (in this example, they are in namespace `ns`, where `person` is defined).
```
Comment (nlohmann) [inline README.md]: Is it possible to define a user type without namespace? If so: please mention this. If not, please describe why.

File: README.md [hunk 8]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+* When using `get<your_type>()`, `your_type` **MUST** be DefaultConstructible and CopyConstructible (There is a way to bypass those requirements described later)
```
Comment (nlohmann) [inline README.md]: Is there a way the feature is useful even if the user-defined type is not DefaultConstructible or CopyConstructible? If so, what can I do if `get()` does not work?

File: README.md [hunk 9]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+  }
+};
+```
+
+This serializer works fine when you have control over the type's namespace.
+However, what about `boost::optional`, or `std::filesystem::path` (C++17)?
+
+Hijacking the `boost` namespace is pretty bad, and it's illegal to add something other than template specializations to `std`...
+
+To solve this, you need to add a specialization of `adl_serializer` to the `nlohmann` namespace, here's an example:
```
Comment (nlohmann) [inline README.md]: If the example works out of the box, please mention this, e.g. "here is a fully-working example", maybe even listing the `#include`s.

File: README.md (hunk 8 above)
Comment (nlohmann) [inline README.md]: As I am reading this top to bottom, I now see that you have a section on this further down. Maybe mention this here as well.

File: README.md [hunk 10]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+  {
+    j = t.i;
+  }
+};
+}
+```
+
+#### Can I write my own serializer? (Advanced use)
+
+Yes. You might want to take a look at `unit-udt.cpp` in the test suite, to see a few examples.
```
Comment (nlohmann) [inline README.md]: Please add a link to the unit test.

File: README.md [hunk 11]
```diff
@@ -440,6 +441,218 @@ int vi = jn.get<int>();
...
+
+If you write your own serializer, you'll need to do a few things:
+
+* use a different `basic_json` alias than nlohmann::json (the last template parameter of basic_json is the JSONSerializer)
```
Comment (nlohmann) [inline README.md]: Is this the `Json` (capital j) below?

File: src/json.hpp.re2c [hunk 12]
```diff
@@ -132,15 +159,156 @@ using is_unscoped_enum =
 
 namespace detail
 {
-// very useful construct against boilerplate (more boilerplate needed than in
-// C++17: http://en.cppreference.com/w/cpp/types/void_t)
-template <typename...> struct make_void
+template <typename Json> std::string type_name(const  Json& j)
```
Comment (nlohmann) [inline src/json.hpp.re2c]: I don't really like the name `Json` here, because it is so easy to confuse with `json`. I have not found a better name though.

File: src/json.hpp.re2c [hunk 13]
```diff
@@ -132,15 +159,156 @@ using is_unscoped_enum =
...
+        j.m_type = value_t::object;
+        j.m_value.object =
+            j.template create<typename Json::object_t>(begin(obj), end(obj));
+        j.assert_invariant();
+    }
 };
-template <typename... Ts> using void_t = typename make_void<Ts...>::type;
 
-// Implementation of 3 C++17 constructs: conjunction, disjunction, negation.
+// Implementation of 2 C++17 constructs: conjunction, negation.
```
Comment (nlohmann) [inline src/json.hpp.re2c]: Can you move these to the top so they do not stand between the constructors?

File: src/json.hpp.re2c [hunk 14]
```diff
@@ -205,50 +368,40 @@ struct is_compatible_object_type_impl<true, RealType, CompatibleObjectType>
         typename CompatibleObjectType::mapped_type>::value;
 };
 
-template<class RealType, class CompatibleObjectType>
+template<class BasicJson, class CompatibleObjectType>
```
Comment (nlohmann) [inline src/json.hpp.re2c]: Same as above: `BasicJson` is really easy to get wrong with `basic_json`.

**Comment by nlohmann** [comment]
Can you please update the PR to the latest develop version? Fixing #296 changed quite some code, though most is to replace `try`/`catch`/`throw` with the macros `JSON_TRY()`/`JSON_CATCH()`/`JSON_THROW()`.

File: src/json.hpp.re2c [hunk 15]
```diff
@@ -354,39 +476,364 @@ struct has_to_json
...
+
+template <typename Json, typename ArithmeticType,
+          enable_if_t<std::is_arithmetic<ArithmeticType>::value and
+                      not std::is_same<ArithmeticType,
+                                       typename Json::boolean_t>::value,
+                      int> = 0>
+void get_arithmetic_value(const  Json& j, ArithmeticType& val)
+{
+    // unsigned must be checked first, since is_number_integer() == true for unsigned
+    if (j.is_number_unsigned())
```
Comment (nlohmann) [inline src/json.hpp.re2c]: I think this may be clearer with a `switch`.

File: src/json.hpp.re2c [hunk 16]
```diff
@@ -354,39 +476,364 @@ struct has_to_json
...
+    enable_if_t <
+        std::is_arithmetic<ArithmeticType>::value and
+        not std::is_same<ArithmeticType, typename Json::number_unsigned_t>::value and
+        not std::is_same<ArithmeticType, typename Json::number_integer_t>::value and
+        not std::is_same<ArithmeticType, typename Json::number_float_t>::value and
+        not std::is_same<ArithmeticType, typename Json::boolean_t>::value,
+        int > = 0 >
+void from_json(const  Json& j, ArithmeticType& val)
+{
+    if (j.is_number_unsigned())
```
Comment (nlohmann) [inline src/json.hpp.re2c]: A `switch` should be clearer.

**Comment by theodelrieu** [comment]
@nlohmann Hi, I rebased my branch onto develop, it would be cool to `reset --hard develop` on the remote too. All the commits that are already in feature/fromtojson were kept (or squashed for some)

File: src/json.hpp.re2c (hunk 12 above)
Comment (theodelrieu) [inline src/json.hpp.re2c]: `Json` is just like `basic_json<Args...>` (which cannot be written, unfortunately, since it has a lot of template template parameters)

I would like to find a good name too, I'll think about it

File: src/json.hpp.re2c (hunk 13 above)
Comment (theodelrieu) [inline src/json.hpp.re2c]: sure

File: src/json.hpp.re2c (hunk 14 above)
Comment (theodelrieu) [inline src/json.hpp.re2c]: Yes, the worst part is that `BasicJson == Json` ... I'll rename it to `Json`, while trying to find a better name for the latter

File: README.md (hunk 7 above)
Comment (theodelrieu) [inline README.md]: Yes if both are in the global namespace, I'll try to phrase that well

**Comment by theodelrieu** [comment]
Unless I missed something, it is ready to be merged (minus documentation).
@nlohmann I kept all the `primitive_iterator_t` methods in the end, they are needed to avoid the **implicit conversion madness**

**Comment by theodelrieu** [comment]
@nlohmann Code-wise, I'm done! Ready to merge :)

**Comment by nlohmann** [comment]
Very nice! I see whether I can do the merge tomorrow.

**Comment by nlohmann** [comment]
Changes moved to #435 which will be merged once the tests complete. Thanks everybody for the hard work over the last months!
